capten/agent/pkg/capten-store/cloud_provider.go

```diff
@@ -0,0 +1,136 @@
+)
+
+func (a *Store) UpsertCloudProvider(config *captenpluginspb.CloudProvider) error {
+	config.LastUpdateTime = time.Now().Format(time.RFC3339)
+	kvPairs, isEmptyUpdate := formUpdateKvPairsForCloudProvider(config)
+	batch := a.client.Session().NewBatch(gocql.LoggedBatch)
+	batch.Query(fmt.Sprintf(insertCloudProviderId, a.keyspace), config.Id)
+	if !isEmptyUpdate {
+		batch.Query(fmt.Sprintf(updateCloudProviderById, a.keyspace, kvPairs), config.Id)
+	}
+	return a.client.Session().ExecuteBatch(batch)
+}
+
+func (a *Store) DeleteCloudProviderById(id string) error {
+	deleteAction := a.client.Session().Query(fmt.Sprintf(deleteCloudProviderById,
+		a.keyspace), id)
+	err := deleteAction.Exec()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *Store) GetCloudProviderForID(id string) (*captenpluginspb.CloudProvider, error) {
+	query := fmt.Sprintf(selectGetCloudProviderById, a.keyspace, id)
+	projects, err := a.executeCloudProvidersSelectQuery(query)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(projects) != 1 {
+		return nil, fmt.Errorf("project not found")
+	}
+	return projects[0], nil
+}
+
+func (a *Store) GetCloudProviders() ([]*captenpluginspb.CloudProvider, error) {
+	query := fmt.Sprintf(selectAllCloudProviders, a.keyspace)
+	return a.executeCloudProvidersSelectQuery(query)
+}
+
+func (a *Store) GetCloudProvidersByLabels(searchLabels []string) ([]*captenpluginspb.CloudProvider, error) {
+	if len(searchLabels) == 0 {
+		return nil, fmt.Errorf("searchLabels empty")
+	}
+
+	labelContains := []string{}
+	for _, label := range searchLabels {
+		labelContains = append(labelContains, fmt.Sprintf("labels CONTAINS '%s'", label))
+	}
+	whereLabelsClause := strings.Join(labelContains, " OR ")
+	whereLabelsClause += " ALLOW FILTERING"
+	query := fmt.Sprintf(selectAllCloudProvidersByLabels, a.keyspace, whereLabelsClause)
+	return a.executeCloudProvidersSelectQuery(query)
+}
+
+func (a *Store) executeCloudProvidersSelectQuery(query string) ([]*captenpluginspb.CloudProvider, error) {
+	selectQuery := a.client.Session().Query(query)
+	iter := selectQuery.Iter()
+
+	provider := captenpluginspb.CloudProvider{}
+	var labels []string
+
+	ret := make([]*captenpluginspb.CloudProvider, 0)
+	for iter.Scan(
+		&provider.Id, &provider.CloudType,
+		&labels, &provider.LastUpdateTime,
+	) {
+		labelsTmp := make([]string, len(labels))
+		copy(labelsTmp, labels)
+		CloudProvider := &captenpluginspb.CloudProvider{
+			Id:             provider.Id,
+			CloudType:      provider.CloudType,
+			Labels:         labelsTmp,
+			LastUpdateTime: provider.LastUpdateTime,
+		}
+		ret = append(ret, CloudProvider)
+	}
+
+	if err := iter.Close(); err != nil {
+		return nil, errors.WithMessage(err, "failed to iterate through results:")
+	}
+
+	return ret, nil
+}
+
+func formUpdateKvPairsForCloudProvider(config *captenpluginspb.CloudProvider) (kvPairs string, emptyUpdate bool) {
```

Comment:
add or update send all the attributes, it shouldn't basis of missing attributes for update.
it need to be inserted, if exist then do update